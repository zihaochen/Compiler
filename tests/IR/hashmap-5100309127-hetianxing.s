.data
	_printf_cnt: .space 4
	_hashsize: .space 4
	_table: .space 4
	ASU_1: .space 404
	str_1: .space 28
.text
main:
	move $fp, $sp     # start using memory here
	move $gp, $sp     # set global pointer (unused)
	subu $sp, $sp, 4
	la $t0, ASU_1
	sw $t0, _table
	li $t0, 0
	sw $t0, _printf_cnt
	li $t0, 100
	sw $t0, _hashsize
	li $t0, 37
	la $t1, str_1
	sw $t0, 0($t1)
	li $t0, 100
	la $t1, str_1
	sw $t0, 4($t1)
	li $t0, 32
	la $t1, str_1
	sw $t0, 8($t1)
	li $t0, 37
	la $t1, str_1
	sw $t0, 12($t1)
	li $t0, 100
	la $t1, str_1
	sw $t0, 16($t1)
	li $t0, 10
	la $t1, str_1
	sw $t0, 20($t1)
	j _main
_getHash:
	sw $ra, ($sp)
	li $t2, 237
	lw $t1, 4($sp)
	mul $t0, $t1, $t2
	sw $t0, -4($sp)
	lw $t2, _hashsize
	lw $t1, -4($sp)
	rem $t0, $t1, $t2
	sw $t0, -8($sp)
	lw $v0, -8($sp)
	sw $v0, -8($sp)
	jr $ra
	jr $ra
_put:
	sw $ra, ($sp)
	li $t0, 0
	sw $t0, -8($sp)
	lw $t0, 8($sp)
	sw $t0, -252($sp)
	subu $sp, $sp, 256
	jal _getHash
	addi $sp, $sp, 256
	sw $v0, -16($sp)
	lw $ra, ($sp)
	lw $t0, -16($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -12($sp)
	li $t0, 0
	sw $t0, -20($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -24($sp)
	lw $t2, -24($sp)
	lw $t1, -20($sp)
	add $t0, $t1, $t2
	sw $t0, -20($sp)
	lw $t2, -20($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t0, -28($sp)
	lw $t1, ($t0)
	sw $t1, -32($sp)
	li $t2, 0
	lw $t1, -32($sp)
	seq $t0, $t1, $t2
	sw $t0, -36($sp)
	lw $t0, -36($sp)
	bne $t0, $0, L2
	b L1
L2:
	li $t0, 16
	sw $t0, -252($sp)
	subu $sp, $sp, 256
	jal _malloc_struct
	addi $sp, $sp, 256
	sw $v0, -48($sp)
	lw $ra, ($sp)
	li $t0, 0
	sw $t0, -52($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -56($sp)
	lw $t2, -56($sp)
	lw $t1, -52($sp)
	add $t0, $t1, $t2
	sw $t0, -52($sp)
	lw $t2, -52($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -60($sp)
	lw $t0, -60($sp)
	lw $t1, -48($sp)
	sw $t1, ($t0)
	lw $t0, -60($sp)
	lw $t1, ($t0)
	sw $t1, -40($sp)
	lw $t0, -60($sp)
	lw $t1, ($t0)
	sw $t1, -40($sp)
	li $t0, 0
	sw $t0, -72($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -76($sp)
	lw $t2, -76($sp)
	lw $t1, -72($sp)
	add $t0, $t1, $t2
	sw $t0, -72($sp)
	lw $t2, -72($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -80($sp)
	lw $t0, -80($sp)
	lw $t1, ($t0)
	sw $t1, -84($sp)
	lw $t0, -84($sp)
	lw $t1, ($t0)
	sw $t1, -88($sp)
	lw $t1, -88($sp)
	lw $t2, 8($sp)
	sw $t2, 0($t1)
	lw $t1, -88($sp)
	lw $t0, 0($t1)
	sw $t0, -68($sp)
	li $t0, 0
	sw $t0, -96($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -100($sp)
	lw $t2, -100($sp)
	lw $t1, -96($sp)
	add $t0, $t1, $t2
	sw $t0, -96($sp)
	lw $t2, -96($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -104($sp)
	lw $t0, -104($sp)
	lw $t1, ($t0)
	sw $t1, -108($sp)
	lw $t0, -108($sp)
	lw $t1, ($t0)
	sw $t1, -112($sp)
	lw $t1, -112($sp)
	lw $t2, 4($sp)
	sw $t2, 4($t1)
	lw $t1, -112($sp)
	lw $t0, 4($t1)
	sw $t0, -92($sp)
	li $t0, 0
	sw $t0, -120($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -124($sp)
	lw $t2, -124($sp)
	lw $t1, -120($sp)
	add $t0, $t1, $t2
	sw $t0, -120($sp)
	lw $t2, -120($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -128($sp)
	lw $t0, -128($sp)
	lw $t1, ($t0)
	sw $t1, -132($sp)
	lw $t0, -132($sp)
	lw $t1, ($t0)
	sw $t1, -136($sp)
	lw $t1, -136($sp)
	li $t2, 0
	sw $t2, 8($t1)
	lw $t1, -136($sp)
	lw $t0, 8($t1)
	sw $t0, -116($sp)
	jr $ra
L1:
	li $t0, 0
	sw $t0, -144($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -148($sp)
	lw $t2, -148($sp)
	lw $t1, -144($sp)
	add $t0, $t1, $t2
	sw $t0, -144($sp)
	lw $t2, -144($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -152($sp)
	lw $t0, -152($sp)
	lw $t1, ($t0)
	sw $t1, -156($sp)
	lw $t0, -156($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -140($sp)
L4:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -160($sp)
	lw $t1, -160($sp)
	lw $t0, 0($t1)
	sw $t0, -164($sp)
	lw $t2, 8($sp)
	lw $t1, -164($sp)
	sne $t0, $t1, $t2
	sw $t0, -168($sp)
	lw $t0, -168($sp)
	bne $t0, $0, L5
	b L3
L5:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -172($sp)
	lw $t1, -172($sp)
	lw $t0, 8($t1)
	sw $t0, -176($sp)
	li $t2, 0
	lw $t1, -176($sp)
	seq $t0, $t1, $t2
	sw $t0, -180($sp)
	lw $t0, -180($sp)
	bne $t0, $0, L7
	b L6
L7:
	li $t0, 16
	sw $t0, -252($sp)
	subu $sp, $sp, 256
	jal _malloc_struct
	addi $sp, $sp, 256
	sw $v0, -192($sp)
	lw $ra, ($sp)
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -196($sp)
	lw $t1, -196($sp)
	lw $t2, -192($sp)
	sw $t2, 8($t1)
	lw $t1, -196($sp)
	lw $t0, 8($t1)
	sw $t0, -184($sp)
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -204($sp)
	lw $t1, -204($sp)
	lw $t0, 8($t1)
	sw $t0, -208($sp)
	lw $t0, -208($sp)
	lw $t1, ($t0)
	sw $t1, -212($sp)
	lw $t1, -212($sp)
	lw $t2, 8($sp)
	sw $t2, 0($t1)
	lw $t1, -212($sp)
	lw $t0, 0($t1)
	sw $t0, -200($sp)
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -220($sp)
	lw $t1, -220($sp)
	lw $t0, 8($t1)
	sw $t0, -224($sp)
	lw $t0, -224($sp)
	lw $t1, ($t0)
	sw $t1, -228($sp)
	lw $t1, -228($sp)
	li $t2, 0
	sw $t2, 8($t1)
	lw $t1, -228($sp)
	lw $t0, 8($t1)
	sw $t0, -216($sp)
L6:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -236($sp)
	lw $t1, -236($sp)
	lw $t0, 8($t1)
	sw $t0, -240($sp)
	lw $t0, -240($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -232($sp)
	b L4
L3:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -248($sp)
	lw $t1, -248($sp)
	lw $t2, 4($sp)
	sw $t2, 4($t1)
	lw $t1, -248($sp)
	lw $t0, 4($t1)
	sw $t0, -244($sp)
	jr $ra
_get:
	sw $ra, ($sp)
	li $t0, 0
	sw $t0, -8($sp)
	lw $t0, 4($sp)
	sw $t0, -68($sp)
	subu $sp, $sp, 72
	jal _getHash
	addi $sp, $sp, 72
	sw $v0, -16($sp)
	lw $ra, ($sp)
	li $t0, 0
	sw $t0, -20($sp)
	li $t2, 4
	lw $t1, -16($sp)
	mul $t0, $t1, $t2
	sw $t0, -24($sp)
	lw $t2, -24($sp)
	lw $t1, -20($sp)
	add $t0, $t1, $t2
	sw $t0, -20($sp)
	lw $t2, -20($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t0, -28($sp)
	lw $t1, ($t0)
	sw $t1, -32($sp)
	lw $t0, -32($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -12($sp)
L9:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -36($sp)
	lw $t1, -36($sp)
	lw $t0, 0($t1)
	sw $t0, -40($sp)
	lw $t2, 4($sp)
	lw $t1, -40($sp)
	sne $t0, $t1, $t2
	sw $t0, -44($sp)
	lw $t0, -44($sp)
	bne $t0, $0, L10
	b L8
L10:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -52($sp)
	lw $t1, -52($sp)
	lw $t0, 8($t1)
	sw $t0, -56($sp)
	lw $t0, -56($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -48($sp)
	b L9
L8:
	lw $t0, -8($sp)
	lw $t1, ($t0)
	sw $t1, -60($sp)
	lw $t1, -60($sp)
	lw $t0, 4($t1)
	sw $t0, -64($sp)
	lw $v0, -64($sp)
	sw $v0, -64($sp)
	jr $ra
	jr $ra
_main:
	sw $ra, ($sp)
	la $t0, str_1
	sw $t0, -80($sp)
	li $t0, 0
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -8($sp)
L12:
	lw $t2, _hashsize
	lw $t1, -4($sp)
	slt $t0, $t1, $t2
	sw $t0, -12($sp)
	lw $t0, -12($sp)
	bne $t0, $0, L13
	b L11
L13:
	li $t0, 0
	sw $t0, -20($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -24($sp)
	lw $t2, -24($sp)
	lw $t1, -20($sp)
	add $t0, $t1, $t2
	sw $t0, -20($sp)
	lw $t2, -20($sp)
	lw $t1, _table
	add $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t0, -28($sp)
	li $t1, 0
	sw $t1, ($t0)
	lw $t0, -28($sp)
	lw $t1, ($t0)
	sw $t1, -16($sp)
	lw $t0, -28($sp)
	lw $t1, ($t0)
	sw $t1, -16($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -44($sp)
	lw $t0, -44($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -40($sp)
	li $t2, 1
	lw $t1, -40($sp)
	sub $t0, $t1, $t2
	sw $t0, -36($sp)
	b L12
L11:
	li $t0, 0
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -48($sp)
L15:
	li $t2, 1000
	lw $t1, -4($sp)
	slt $t0, $t1, $t2
	sw $t0, -52($sp)
	lw $t0, -52($sp)
	bne $t0, $0, L16
	b L14
L16:
	lw $t0, -4($sp)
	sw $t0, -104($sp)
	lw $t0, -4($sp)
	sw $t0, -108($sp)
	subu $sp, $sp, 112
	jal _put
	addi $sp, $sp, 112
	sw $v0, -56($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -68($sp)
	lw $t0, -68($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -64($sp)
	li $t2, 1
	lw $t1, -64($sp)
	sub $t0, $t1, $t2
	sw $t0, -60($sp)
	b L15
L14:
	li $t0, 0
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -72($sp)
L18:
	li $t2, 1000
	lw $t1, -4($sp)
	slt $t0, $t1, $t2
	sw $t0, -76($sp)
	lw $t0, -76($sp)
	bne $t0, $0, L19
	b L17
L19:
	lw $t0, -4($sp)
	sw $t0, -104($sp)
	subu $sp, $sp, 108
	jal _get
	addi $sp, $sp, 108
	sw $v0, -84($sp)
	lw $ra, ($sp)
	lw $t0, -80($sp)
	sw $t0, -104($sp)
	lw $t0, -4($sp)
	sw $t0, -108($sp)
	lw $t0, -84($sp)
	sw $t0, -112($sp)
	li $t0, 3
	sw $t0, _printf_cnt
	subu $sp, $sp, 116
	jal _printf
	addi $sp, $sp, 116
	sw $v0, -88($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -100($sp)
	lw $t0, -100($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -96($sp)
	li $t2, 1
	lw $t1, -96($sp)
	sub $t0, $t1, $t2
	sw $t0, -92($sp)
	b L18
L17:
	jr $ra

_printf:
	lw $t0, _printf_cnt
	mul $t0, $t0, 4
	add $t0, $t0, $sp
	lw $a1, ($t0)
	move $a2, $t0
		
_printf_loop:
	lb $a0, 0($a1)
	beq $a0, 0, _printf_end
	addu $a1, $a1, 4
	beq $a0, '%', _printf_fmt
	li $v0, 11
	syscall
	b _printf_loop

_printf_fmt:
	lb $a0, 0($a1)
	addu $a1, $a1, 4
	beq $a0, 'd', _printf_int
	beq $a0, 's', _printf_str
	beq $a0, 'c', _printf_char
	beq $a0, '0', _printf_width
	beq $a0, '.', _printf_width

_printf_int:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 1
	syscall
	b _printf_loop

_printf_str:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 4
	syscall
	b _printf_loop

_printf_char:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 11
	syscall
	b _printf_loop

_printf_width:
	lb $t1, 0($a1)
	addu $a1, $a1, 4
	sub $t1, $t1, '1'
	addu $a1, $a1, 4
	subu $a2, $a2, 4
	lw $t0, 0($a2)
	blt $t0, 10, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000000, _printf_width_tmp
	subu $t1, $t1, 1

_printf_width_tmp:
	li $a0, '0'
	beq $t1, 0, _printf_width_tmp2
	subu $t1, $t1, 1
	li $v0, 11
	syscall
	b _printf_width_tmp

_printf_width_tmp2:
	move $a0, $t0
	li $v0, 1
	syscall
	b _printf_loop

_printf_end:
	jr $ra

_getchar:
	li $v0, 12
	syscall
	jr $ra

_malloc_struct:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	move $t0, $v0
	li $v0, 9
	li $a0, 4
	syscall
	sw $t0, ($v0)
	jr $ra

_malloc:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	jr $ra

