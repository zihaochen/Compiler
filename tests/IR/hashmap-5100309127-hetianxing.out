hashsize = 100
Temp1031 = n MUL 237
Temp1032 = Temp1031 MOD hashsize
Return BasicParam Temp1032

ptr = 0
BasicParam key
call getHash , 1
p = Temp1034
Temp1033 = p
Temp1035 = 0
Temp1036 = p MUL 12
Temp1035 = Temp1035 ADD Temp1036
Temp1037 = table ADD Temp1035
Temp1038 = Temp1037
Temp1039 = Temp1038 EQ 0
If Not Equal to Zero Temp1039 Goto Label :115	
Goto label:116
Label :115	BasicParam 12
call malloc , 1
Temp1042 = 0
Temp1043 = p MUL 12
Temp1042 = Temp1042 ADD Temp1043
Temp1044 = table ADD Temp1042
Temp1045 = Temp1044
Temp1044 = Temp1041
Temp1047 = 0
Temp1048 = p MUL 12
Temp1047 = Temp1047 ADD Temp1048
Temp1049 = table ADD Temp1047
Temp1050 = Temp1049
Temp1051 = *Temp1050, 12
Temp1051[0] = key , 4
Temp1053 = 0
Temp1054 = p MUL 12
Temp1053 = Temp1053 ADD Temp1054
Temp1055 = table ADD Temp1053
Temp1056 = Temp1055
Temp1057 = *Temp1056, 12
Temp1057[4] = data , 4
Temp1059 = 0
Temp1060 = p MUL 12
Temp1059 = Temp1059 ADD Temp1060
Temp1061 = table ADD Temp1059
Temp1062 = Temp1061
Temp1063 = *Temp1062, 12
Temp1063[8] = 0 , 4
Return 
Goto label:114
Label :116	Label :114	Temp1065 = 0
Temp1066 = p MUL 12
Temp1065 = Temp1065 ADD Temp1066
Temp1067 = table ADD Temp1065
Temp1068 = Temp1067
ptr = Temp1068
Temp1064 = ptr
Label :118	Temp1069 = *ptr, 12
Temp1070 = Temp1069[0] , 4
Temp1071 = Temp1070 NE key
If Not Equal to Zero Temp1071 Goto Label :119	
Goto label:117
Label :119	Temp1072 = *ptr, 12
Temp1073 = Temp1072[8] , 4
Temp1074 = Temp1073 EQ 0
If Not Equal to Zero Temp1074 Goto Label :121	
Goto label:122
Label :121	BasicParam 12
call malloc , 1
Temp1077 = *ptr, 12
Temp1077[8] = Temp1076 , 4
Temp1079 = *ptr, 12
Temp1080 = Temp1079[8] , 4
Temp1081 = *Temp1080, 8
Temp1081[0] = key , 4
Temp1083 = *ptr, 12
Temp1084 = Temp1083[8] , 4
Temp1085 = *Temp1084, 8
Temp1085[8] = 0 , 4
Goto label:120
Label :122	Label :120	Temp1087 = *ptr, 12
Temp1088 = Temp1087[8] , 4
ptr = Temp1088
Temp1086 = ptr
Goto label:118
Label :117	Temp1090 = *ptr, 12
Temp1090[4] = data , 4
ptr = 0
BasicParam key
call getHash , 1
Temp1093 = 0
Temp1094 = Temp1092 MUL 12
Temp1093 = Temp1093 ADD Temp1094
Temp1095 = table ADD Temp1093
Temp1096 = Temp1095
ptr = Temp1096
Temp1091 = ptr
Label :124	Temp1097 = *ptr, 12
Temp1098 = Temp1097[0] , 4
Temp1099 = Temp1098 NE key
If Not Equal to Zero Temp1099 Goto Label :125	
Goto label:123
Label :125	Temp1101 = *ptr, 12
Temp1102 = Temp1101[8] , 4
ptr = Temp1102
Temp1100 = ptr
Goto label:124
Label :123	Temp1103 = *ptr, 12
Temp1104 = Temp1103[4] , 4
Return BasicParam Temp1104

i = 0
Temp1105 = i
Label :127	Temp1106 = i LT hashsize
If Not Equal to Zero Temp1106 Goto Label :128	
Goto label:126
Label :128	Temp1108 = 0
Temp1109 = i MUL 12
Temp1108 = Temp1108 ADD Temp1109
Temp1110 = table ADD Temp1108
Temp1111 = Temp1110
Temp1110 = 0
Temp1112 = i
i = i ADD 1
Goto label:127
Label :126	i = 0
Temp1113 = i
Label :130	Temp1114 = i LT 1000
If Not Equal to Zero Temp1114 Goto Label :131	
Goto label:129
Label :131	BasicParam i
BasicParam i
call put , 2
Temp1116 = i
i = i ADD 1
Goto label:130
Label :129	i = 0
Temp1117 = i
Label :133	Temp1118 = i LT 1000
If Not Equal to Zero Temp1118 Goto Label :134	
Goto label:132
Label :134	Temp1119 = &"%d %d
"
BasicParam i
call get , 1
BasicParam Temp1119
BasicParam i
BasicParam Temp1120
call printf , 3
Temp1122 = i
i = i ADD 1
Goto label:133
Label :132	